### PR TITLE
planner: support trace topn push down

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/29661

### What is changed and how it works?

support trace topn push down

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
None
```

### CODE REVIEW

**Review by zz-jason — COMMENTED** [review]
There are lots of `fmt.Sprintf` in the critical paths. It's better to only trigger these evaluations when optimizer tracing is turned on to save the performance. Maybe you can try `fmt.Stringer` just like we did in collecting metrics.

File: planner/core/rule_topn_push_down.go [hunk 1]
```diff
@@ -62,57 +64,67 @@ func (lt *LogicalTopN) setChild(p LogicalPlan) LogicalPlan {
...
 	// If a TopN is pushed down, this sort is useless.
-	return ls.children[0].pushDownTopN(topN)
+	return ls.children[0].pushDownTopN(topN, opt)
 }
 
-func (p *LogicalLimit) convertToTopN() *LogicalTopN {
-	return LogicalTopN{Offset: p.Offset, Count: p.Count, limitHints: p.limitHints}.Init(p.ctx, p.blockOffset)
+func (p *LogicalLimit) convertToTopN(opt *logicalOptimizeOp) *LogicalTopN {
+	topn := LogicalTopN{Offset: p.Offset, Count: p.Count, limitHints: p.limitHints}.Init(p.ctx, p.blockOffset)
+	opt.appendStepToCurrent(topn.ID(), topn.TP(), "Limit can be converted into TopN", fmt.Sprintf("%v[%v] is converted into %v[%v]",
```
Comment (zz-jason) [inline planner/core/rule_topn_push_down.go:90]: through the `%v[%v]` format looks more structured to describe operators, it's more important to keep consistent with the format `%v_%v` used by the result of `explain`.

One of the typical scenario of using optimizer tracing is:
- execute `explain` to see the execution plan of a slow query
- execute the optimizer tracing to see the internals of why the execution plan is generated, how each operator in the plan is generated

Comment (Yisaer) [reply planner/core/rule_topn_push_down.go]: updated

**Review by rebelice — APPROVED** [review]

**Comment by Yisaer** [comment]
> Stringer

good catch, lazy initial output will be done in https://github.com/pingcap/tidb/pull/30882

**Comment by Yisaer** [comment]
> There are lots of `fmt.Sprintf` in the critical paths. It's better to only trigger these evaluations when optimizer tracing is turned on to save the performance. Maybe you can try `fmt.Stringer` just like we did in collecting metrics.

good catch, lazy initial output will be done in https://github.com/pingcap/tidb/pull/30882

**Review by JmPotato — APPROVED** [review]

**Pull request merged**